aip/general/0180.md

```diff
@@ -139,6 +139,84 @@ code, as such changes will be seen as breaking by those users.
 whether a proposed change is likely to break users, and an expansive reading of
 this guidance could ostensibly prevent _any_ change (which is not the intent).
 
+#### Default values must not change
+
+Changing the default value is considered breaking and **must not** be done. The
+default behavior for a resource is determined by it's default values, and this
+**must not** change across minor versions.
+
+For example:
+
+```proto
+// A representation of an automobile
+message Automobile {
+    // The name of the automobile.
+    string name = 1;
+    // The number of wheels of the automobile.
+    // The default is set to 2 when no value is sent by the client
+    int wheels = 2;
+}
+```
+
+Changing to:
+
+```proto
+// A representation of an automobile
+message Automobile {
+    // The name of the automobile.
+    string name = 1;
+    // The number of wheels of the automobile.
+    // The default is set to 4 when no value is sent by the client
+    int wheels = 2;
+}
+```
+
+would constitute a breaking change.
+
+#### Serializing defaults
+
+APIs **must not** change the way a field with a default value is serialized. For
```

Comment:
This seems reasonable - I suspect it'll all get muddier when we've thought more about field presence, but I'm okay with this description as it is for now.

Changing it to "resource field" might be worth while though.

Reply:
A lot of the difficulty here comes from inconsistent behavior for serializing a default value (default as in runtime-set that is not defined in the proto) so I actually think that field presence would fix it. In a world where field presence is meaningful the problem goes away because it's clear that what unset means.

Reply:
And you can't have default values for fields where presence is meaningful. At least that's my understanding

Reply:
Hmm... I can imagine a case where field presence is useful in create/update (so that the client can say whether they're trying to update it) but where you *always* want the server to specify a value when fetching, so the consumer can know what value is in effect. That makes it sort of "semi-optional" - which can't easily be expressed just in proto terms, of course.

Reply:
If field presence is meaningful in create/update then declarative clients need to know what value was actually set (or unset) when they retrieve the resource. "semi-optional" is not a feasible solution, and must be prohibited from existing.

Reply:
If we do require this type of functionality it *needs* to be split between separate fields for the input value and the effective value

Reply:
> If field presence is meaningful in create/update then declarative clients need to know what value was actually set (or unset) when they retrieve the resource. "semi-optional" is not a feasible solution, and must be prohibited from existing.

Can you go into more detail about this? Suppose we have a situation where the client doesn't specify a value, and the server populates it with a default and returns that value. So long as the result is *exactly* the same as if the client specified that value to start with, is there still a problem? I may well be missing something. If you could give an example of a problematic flow, I'll see whether we're talking about different things, or whether there's an aspect I was missing.

Reply:
To be clear: I still agree with the paragraph - the server shouldn't suddenly start or stop populating it. It was the side-issue of "semi-optional".

Reply:
> Can you go into more detail about this? Suppose we have a situation where the client doesn't specify a value, and the server populates it with a default and returns that value. So long as the result is exactly the same as if the client specified that value to start with, is there still a problem? I may well be missing something. If you could give an example of a problematic flow, I'll see whether we're talking about different things, or whether there's an aspect I was missing.

My understanding of this "semi-optional" field is as such:

If the field is specified by the client it works as normal.
If the field is unspecified by the client, a server-set default is returned

There are a couple potential behaviors for when a client sends an empty value during a PATCH request:

1. Server ignores the field
2. Server sets the field to the empty value
3. Server sets the field to the default value it would have chosen if the resource was created with the field unspecified

1 is just a normal field where presence is not meaningful
2 is inconsistent with the behavior of the field on create, which doesn't make sense. In order to create a resource with an empty value for this field the client would need to create and then update immediately to remove the field value
3 is implementing a sort of "reset" functionality for the field, which doesn't seem to fit standard CRUD. It could make sense to have on a custom method if the functionality is needed for the API, but doesn't fit into standard resource behavior that should be implemented in clients

Reply:
1 isn't quite the same as a normal field - because if the field has the `optional` keyword (i.e. presence is available) then the server can tell the difference between the client sending a value of 0 and not sending anything at all. I think it would be reasonable to treat not sending anything as "don't change this", if the field is always *effectively* present.

(There's a harder decision around how to send an update which explicitly wants to *clear* a field when that's a meaningful state.)

Reply:
Oh, yeah that's an odd one. It's similar to a normal field with a server-set default, but adds the zero value as a settable value. I'm concerned with how we would communicate this to a client, as the `optional` keyword will generally signify that a field can be unset, but in this situation sending unset would have an entirely different behavior.

What would the overlap be between the unset value meaning "don't change" and setting an updateMask for that field? What if these conflict, where a user sends unset but sets the updateMask

Reply:
> What would the overlap be between the unset value meaning "don't change" and setting an updateMask for that field?

Not as well defined as I'd like, basically. (That tends to be the case for anything to do with FieldMask, unfortunately.)

> What if these conflict, where a user sends unset but sets the updateMask

This *could* be the signal used for "please clear the field" - which would be invalid in the case I'm thinking of, where the field always *must* have a value, it's just it doesn't have to be specified by the user.

Taking a step back, "The user doesn't have to set this explicitly, but 0 is a valid value" sounds like a reasonable pair of requirements to me. It's unclear how *important* it is as a pair of requirements, mind you.

Reply:
> Taking a step back, "The user doesn't have to set this explicitly, but 0 is a valid value" sounds like a reasonable pair of requirements to me.

Yes I agree this is a reasonable behavior for a field. I think the problem is that we don't have a clear way to handle these types of fields currently. We would need a revamp of FieldMask & existing functionality before this can be widely adopted